### PR TITLE
Fix token endpoints ignore specified account

### LINKED ISSUE #26234: Some endpoints ignore specified account and use auth account instead
### Description

Reproduction:
- I created an access token at the `/api/v1/users/bob/tokens` endpoint, using basic auth with my `admin` account.
- I expected the access token to be created for `bob`, but it was actually created for `admin`.
- the same is true for other related endpoints

This was **very** surprising and took a while to debug. Is it a bug or by design?

If it's a bug (I think it is), then this issue will track it.

If it's by design:
- It's common to use an admin account to interact with an app, even when managing other users' accounts. So this behaviour is very surprising.
- I specified that I wanted to act on the `bob` account, not the `admin` account.
- If this cannot be fixed/changed, then at least those API endpoints should return errors. If it will ignore the account that I specified and use the authenticating account instead, it must return an error rather than perform an action I didn't request.
- The docs should reflect this weird behaviour (I looked but couldn't find anything).

Thanks!

### Gitea Version

1.20.2

### Can you reproduce the bug on the Gitea demo site?

No

### Log Gist

n/a

### Screenshots

n/a

### Git Version

n/a

### Operating System

n/a

### How are you running Gitea?

docker

### Database

None

### PR DESCRIPTION
**Pull request by CaiCandong**
Fix #26234
close #26323
close #27040

### CODE REVIEW

**Comment by CaiCandong** [comment]
> #27040 is also related

I made the same change as #27040 in #26323, but the intent of the person who raised the issue was for admins to manage other users' tokens.
So I'm re-initiating this PR.

**Comment by JakobDev** [comment]
If a endpoint is mean to uses by admins, it should be in the `/admins` route. You should also note that admins can use the [sudo](baf44a90136457d6eff6ffdf1cdfc45d3eb8c055) function of the API.

**Comment by CaiCandong** [comment]
> If a endpoint is mean to uses by admins, it should be in the `/admins` route. You should also note that admins can use the [sudo](https://docs.gitea.com/development/api-usage#sudo) function of the API.

Yes, that's how I implemented it in #26323 as well. but

https://github.com/go-gitea/gitea/pull/26323#discussion_r1325390384
>  As an admin I would like it more to use this instead of an admin/users endpoint. I know I'm an admin, I don't need every of my actions behind admin/.

**Comment by JakobDev** [comment]
I don't get that argument. It's nice that you know that, but that doesn't matter. here.  A API should try to keep a consistent design to not confuse users. And the current design is, that all endpoints under `/username/{username}` can be used by any user (some even with login) and that endpoints, who are meant to be used by admins should live under `/admin`. Most of the users of the API are not Admins.

I think the best way to solve this is to simply move the endpoint to `/user/token`. It will only work for the current User. If you are a Admin, you can simply use the sudo parameter/header of the API. It works the same way as editing e.g. OAuth2Application though the API or changing the Avatar.

**Comment by KN4CK3R** [comment]
I we introduce a new role `moderator` which could edit users and repos but not admin related stuff like auth sources, what do we do? Add all routes again under `/moderator`?

For me the `/users/{username}` routes are correct. If you have the permission to perform the action, everything is fine. If you don't have permission, you get the matching error code.

File: routers/api/v1/user/app.go [hunk 1]
```diff
@@ -43,8 +43,14 @@ func ListAccessTokens(ctx *context.APIContext) {
...
 	//   "200":
 	//     "$ref": "#/responses/AccessTokenList"
+	//   "403":
+	//     "$ref": "#/responses/forbidden"
 
-	opts := auth_model.ListAccessTokensOptions{UserID: ctx.Doer.ID, ListOptions: utils.GetListOptions(ctx)}
+	if !ctx.Doer.IsAdmin && ctx.Doer != ctx.ContextUser {
+		ctx.Error(http.StatusForbidden, "", "Only admin or user self can list access tokens")
+		return
+	}
```
Comment (yp05327) [inline routers/api/v1/user/app.go]: This permission check can be a `reqSelf` function as there are some similar `reqXXX` functions here:
https://github.com/go-gitea/gitea/blob/e5ec57cd608a81be6e0a6b897b165e9b322ceab4/routers/api/v1/api.go#L351-L388

Comment (CaiCandong) [reply routers/api/v1/user/app.go]: Thanks very much!

**Comment by JakobDev** [comment]
I'm not saying the current system is perfect, but at least we have a system. And in the current  system we use the `/user` route for User stuff like e.g. editing OAuth2 Apps, GPG keys, avatars, emails, hooks, settings etc. This `/user` (not `/users/<username>`) route is for the current User. If a Admin wants to use those routes, he can use the sudo  functionality. This may have quirks, but this is how it's currently works. I see no reason why a single endpoint should work differently.

**Comment by KN4CK3R** [comment]
> I see no reason why a single endpoint should work differently.

This is the first and only endpoint, so there is no difference to other endpoints.
Gitlab follows the same scheme:
https://docs.gitlab.com/ee/api/users.html#add-ssh-key
> Creates a new key owned by the authenticated user.
`POST /user/keys`
Create new key owned by specified user. Available only for administrator.
`POST /users/:id/keys`

**Comment by CaiCandong** [comment]
@waTeim  @lonix1 You mentioned that you need to use this interface, what do you think?

**Comment by JakobDev** [comment]
> This is the first and only endpoint, so there is no difference to other endpoints.

As I said, there are other endpoints to edit the settings of a User

>  Gitlab follows the same scheme:

This is Gitea, not GitLab

**Comment by KN4CK3R** [comment]
> > This is the first and only endpoint, so there is no difference to other endpoints.
> 
> As I said, there are other endpoints to edit the settings of a User

Yep, in the admin area where they should not be.

> > Gitlab follows the same scheme:
> 
> This is Gitea, not GitLab

They did not make the mistake.

**Comment by KN4CK3R** [comment]
> I'm a strong proponent of always using plural on REST API URLs.

We do that, just not for the `/user/` endpoint which uses the calling user as context.

**Review by KN4CK3R — APPROVED** [review]

**Review by silverwind — APPROVED** [review]

File: tests/integration/api_token_test.go [hunk 2]
```diff
@@ -40,6 +40,29 @@ func TestAPIDeleteMissingToken(t *testing.T) {
 	MakeRequest(t, req, http.StatusNotFound)
 }
 
+// TestAPIGetTokensPermission ensures that Only admin can create token for other users
```
Comment (silverwind) [inline tests/integration/api_token_test.go]:
```suggestion
// TestAPIGetTokensPermission ensures that only admin can create token for other users
```

**Comment by puni9869** [comment]
Please run the `make generate-swagger`

**Comment by CaiCandong** [comment]
> Please run the `make generate-swagger` once again.

No new changes were made to Swagger, and the re-execution has nothing to commit.

**Comment by CaiCandong** [comment]
Do we need to backport to 1.20? @lunny

**Comment by puni9869** [comment]
No I guess

**Pull request merged**